core/src/main/java/org/apache/iceberg/rest/ScanTasksIterable.java

```diff
@@ -0,0 +1,247 @@
+
+      try {
+        while (true) {
+          if (shutdown.get()) {
+            return;
+          }
+
+          String planTask = planTasks.poll();
+          if (planTask == null) {
+            return;
+          }
+
+          processPlanTask(planTask);
+        }
+
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      } catch (Exception e) {
+        throw new RuntimeException("Worker failed processing planTask", e);
+      } finally {
+        int remaining = activeWorkers.decrementAndGet();
+
+        if (remaining == 0 && !planTasks.isEmpty() && !shutdown.get()) {
+          executorService.execute(new PlanTaskWorker());
+        }
+      }
+    }
+
+    private void processPlanTask(String planTask) throws InterruptedException {
+      FetchScanTasksResponse response = fetchScanTasks(planTask);
+
+      if (response.fileScanTasks() != null) {
+        for (FileScanTask task : response.fileScanTasks()) {
+          if (shutdown.get()) {
+            return;
+          }
+          taskQueue.put(task);
+        }
+      }
+
+      if (response.planTasks() != null && !response.planTasks().isEmpty()) {
+        planTasks.addAll(response.planTasks());
+      }
+    }
+
+    private FetchScanTasksResponse fetchScanTasks(String planTask) {
+      FetchScanTasksRequest request = new FetchScanTasksRequest(planTask);
+
+      return client.post(
+          resourcePaths.fetchScanTasks(tableIdentifier),
+          request,
+          FetchScanTasksResponse.class,
+          headers,
+          ErrorHandlers.defaultErrorHandler(),
+          stringStringMap -> {},
+          parserContext);
+    }
+  }
+
+  private class ScanTasksIterator implements CloseableIterator<FileScanTask> {
+    private FileScanTask nextTask = null;
+
+    @Override
+    public boolean hasNext() {
+      if (nextTask != null) {
+        return true;
+      }
+
+      while (true) {
+        if (isDone()) {
+          return false;
+        }
+
+        try {
+          nextTask = taskQueue.poll(QUEUE_POLL_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+
+          if (nextTask != null) {
+            return true;
+          }
+
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          return false;
+        }
+      }
+    }
+
+    @Override
+    public FileScanTask next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException("No more scan tasks available");
+      }
+      FileScanTask result = nextTask;
+      nextTask = null;
+      return result;
+    }
+
+    @Override
+    public void close() throws IOException {
+      shutdown.set(true);
+
+      if (cancellationCallback != null) {
+        try {
+          @SuppressWarnings("unused")
+          Boolean ignored = cancellationCallback.get();
+        } catch (Exception e) {
+          // Ignore cancellation failures
+        }
+      }
+
+      taskQueue.clear();
+      planTasks.clear();
```

Comment:
Minor: If there's still elements in the queue or in the planTasks list, do we want to log how many, at the time of close?  In the case of a limit expressed by a client, the size of the remainder of these structures effectively represents how much "extra" task fetching was done.